staging/src/k8s.io/client-go/util/csaupgrade/upgrade.go

```diff
@@ -0,0 +1,211 @@
+// with the given `ssaManager` for `Apply` operations.
+//
+// This transformation should be performed on an object if it has been previously
+// managed using client-side-apply to prepare it for future use with
+// server-side-apply.
+//
+// Caveats:
+//  1. This operation is not reversible. Information about which fields the client
+//     owned will be lost in this operation.
+//  2. Supports being performed either before or after initial server-side apply.
+//  3. Client-side apply tends to own more fields (including fields that are defaulted),
+//     this will possibly remove this defaults, they will be re-defaulted, that's fine.
+//  4. Care must be taken to not overwrite the managed fields on the server if they
+//     have changed before sending a patch.
+//
+// obj - Target of the operation which has been managed with CSA in the past
+// csaManagerName - Name of FieldManager formerly used for `Update` operations
+// ssaManagerName - Name of FieldManager formerly used for `Apply` operations
+// Returns: a copy of the `obj` paramter with its managed fields modified so that
+// it may be used with server-side apply in the future.
+func UpgradeManagedFields(
+	obj runtime.Object,
+	csaManagerName string,
+	ssaManagerName string,
+) (runtime.Object, error) {
+	accessor, err := meta.Accessor(obj)
+	if err != nil {
+		return nil, fmt.Errorf("error accessing object metadata: %w", err)
+	}
+
+	// Create managed fields clone since we modify the values
+	var managedFields []metav1.ManagedFieldsEntry
+	managedFields = append(managedFields, accessor.GetManagedFields()...)
+
+	// Locate SSA manager
+	ssaManagerIndex, ssaManagerExists := findFirstIndex(managedFields,
+		func(entry metav1.ManagedFieldsEntry) bool {
+			return entry.Manager == ssaManagerName &&
+				entry.Operation == metav1.ManagedFieldsOperationApply &&
+				entry.Subresource == ""
+		})
+
+	if ssaManagerExists {
+		err = unionManagerIntoIndex(managedFields, ssaManagerIndex, csaManagerName)
+		if err != nil {
+			return nil, err
+		}
+
+	} else {
+		// SSA manager does not exist. Find the most recent matching CSA manager,
+		// convert it to an SSA manager.
+		//
+		// (find first index, since managed fields are sorted so that most recent is
+		//  first in the list)
+		csaManagerIndex, csaManagerExists := findFirstIndex(managedFields,
+			func(entry metav1.ManagedFieldsEntry) bool {
+				return entry.Manager == csaManagerName &&
+					entry.Operation == metav1.ManagedFieldsOperationUpdate &&
+					entry.Subresource == ""
+			})
+
+		if !csaManagerExists {
+			// There are no CSA managers that need to be converted. Nothing to do
+			// Return early
+			return obj, nil
+		}
+
+		// Convert CSA manager into SSA manager
+		managedFields[csaManagerIndex].Operation = metav1.ManagedFieldsOperationApply
+		managedFields[csaManagerIndex].Manager = ssaManagerName
+
+		err = unionManagerIntoIndex(managedFields, csaManagerIndex, csaManagerName)
+		if err != nil {
+			return nil, err
+		}
+	}
```

Comment:
Little refactor, tell me what you think:
```suggestion
	// Locate SSA manager
	replaceIndex, managerExists := findFirstIndex(managedFields,
		func(entry metav1.ManagedFieldsEntry) bool {
			return entry.Manager == ssaManagerName &&
				entry.Operation == metav1.ManagedFieldsOperationApply &&
				entry.Subresource == ""
		})

	if !managerExists {
		// SSA manager does not exist. Find the most recent matching CSA manager,
		// convert it to an SSA manager.
		//
		// (find first index, since managed fields are sorted so that most recent is
		//  first in the list)
		replaceIndex, managerExists = findFirstIndex(managedFields,
			func(entry metav1.ManagedFieldsEntry) bool {
				return entry.Manager == csaManagerName &&
					entry.Operation == metav1.ManagedFieldsOperationUpdate &&
					entry.Subresource == ""
			})

		if !managerExists {
			// There are no CSA managers that need to be converted. Nothing to do
			// Return early
			return obj, nil
		}

		// Convert CSA manager into SSA manager
		managedFields[replaceIndex].Operation = metav1.ManagedFieldsOperationApply
		managedFields[replaceIndex].Manager = ssaManagerName
	}
	err = unionManagerIntoIndex(managedFields, replaceIndex, csaManagerName)
	if err != nil {
		return nil, err
	}
```